### PR TITLE
Allow a Resource to explicitly specify a set of other Resources it depends on.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
When writing components, both @swgillespie and i have run into a similar problem.  We have a component that we want others to be able to say they are dependent on.  However, being dependent on the component doesn't ensure that you will actually end up awaiting the inner custom resources that are created.

This is been quite a nuisance in awsinfra, and has required a lot of contortions to workaround.

One solution people have hacked up here is to swap out the 'urn' or 'id' for a ComponentResource with an output that wraps the true id/urn output, but adds in those resources explicitly into it.  A fairly hacky solution.

This PR introduces a new concept at the Resource level.  specifically Resources can now override a 'dependsOn' method that returns a list of resources that should logically also be depended on if someone depends on this resource.

When registering a resource, we now walk and collect the full transitive set of Resources and ensure they're all complete before continuing with the resource we're currently working on.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 PROJECT_NAME := Pulumi SDK
-SUB_PROJECTS := sdk/nodejs sdk/python sdk/go
+SUB_PROJECTS := sdk/nodejs sdk/go
```
Comment (CyrusNajmabadi) [inline Makefile]: will revert this.

**Comment by lukehoban** [comment]
Curious - is there any reason to do this instead of just making a dependency on a component depend on all children of that component?  Seems like this should just be the default behavior instead of making component authors implement a new method?

Would be good to have the general design discussion around options here on https://github.com/pulumi/pulumi/issues/1743. 

Cc also @pgavlin who was nominally going to be tackling a design for that issue in M20.

**Comment by CyrusNajmabadi** [comment]
adding tests now.

**Comment by CyrusNajmabadi** [comment]
> Curious - is there any reason to do this instead of just making a dependency on a component depend on all children of that component? 

How can one get "all children of the component"?  If that's possible, that works for me.

**Comment by pgavlin** [comment]
> How can one get "all children of the component"? If that's possible, that works for me.

You can't, currently. Defining "all children" is sort of problematic without adding some constraints, as children can attach themselves to a parent after that parent has been created (where "created" can be either at the SDK level or at the engine level).

**Comment by CyrusNajmabadi** [comment]
Right.  Ok.  So this is basically an attempt by me for a resource to say "here are the appropriate resources to think of as my 'children' :)  

given that, what's the read on this approach?  Is it terrible?  Great?  Meh?  Should we be solving this some other way entirely?

--

Note; this is really a much larger concern at the awsinfra layer.  I'm making tons of components and i've run into this several times already where a component depends on another component, but things don't dependsOn properly.

I can take the approach of hacking the URN property of the component so that it's not ready until the child components are done.  But it feels pretty wrong. :)

**Comment by CyrusNajmabadi** [comment]
Closing out.  We're going to do this with the approach specified here: https://github.com/pulumi/pulumi/issues/2295

**Pull request closed without merging**